### DISCUSSION TITLE
Params Request: get per name

### DISCUSSION THREAD

**Discussion by jesusvilla** [uNetworking/uWebSockets.js#236] [Q&A] [answered]
First, let me congratulate you for such a powerful framwork, but I have a problem when I create a route with parameters, I think that the right thing (as in other frameworks: expressjs, fastify) would be for an object to return and not do `req.getParameter(0)`.

```javascript
app.get('/candy/:kind', (res, req) => {
  /* Parameters */
  res.end('So you want candy? Have some ' + req.getParameter(0) + '!');
})
```

Should be:

```javascript
app.get('/candy/:kind', (res, req) => {
  /* Parameters */
  res.end('So you want candy? Have some ' + req.params['kind'] + '!');
})
```

Or in any case:

```javascript
app.get('/candy/:kind', (res, req) => {
  /* Parameters */
  res.end('So you want candy? Have some ' + req.getParameterName('kind') + '!');
})
```
The latter to not conflict with the current API.

**Comment by ghost** [comment]
Lookup by index is what this library does. It is not that hard to keep track of what is what, and mixup can happen either way.

**Comment by jesusvilla** [comment]
I say it because it is better to call him by his name than by his position

**Comment by jesusvilla** [comment]
@alexhultman  Consider it please, many of us will thank you, so you will have more people from Express who migrate to uWebSockets.js.

**Comment by ghost** [comment, marked as answer]
I have considered it. Lookup by index is free, lookup by name requires building some kind of hash map or linearly searching the URL multiple times. It is really not that complex to write

let kind = req.getParameter(0) and then use kind as type safe variable. That mapping is free.

**Reply by yamiteru** [reply]
Can we at least get an array of parameters so we can do something like this?
```typescript
app.get("/article/:author/:slug", (res, req) => {
  const [ author, article ] = req.getParameters();
});
```

**Reply by a-tonchev** [reply]
> Can we at least get an array of parameters so we can do something like this?
> 
> ```ts
> app.get("/article/:author/:slug", (res, req) => {
>   const [ author, article ] = req.getParameters();
> });
> ```

You can do this very easy:
```
app.get("/article/:author/:slug", (res, req) => {
 const [ author, article ] = (new Array(2)).map((el,index) => req.getParameter(index))
});
```

Or just:

```
app.get("/article/:author/:slug", (res, req) => {
 const [ author, article ] = [req.getParameter(0), req.getParameter(1)];
});
```

**Reply by jesusvilla** [reply]
If uws.HttpRequest could be accessed, the .getParameters() method could be added, but this doesn't exist, so it should be like this:
```js
const uws = require('uWebSockets.js');

const app = uws.App({});

const extendHttpRequest = (req) => {
  if (req.constructor._extended !== true) {
    const HttpRequest = req.constructor;
    HttpRequest._extended = true;

    // Extend uws.HttpRequest
    HttpRequest.prototype.getParameters = function () {
      const params = [];
      let param, index = 0;
      while (param = this.getParameter(index)) {
        params.push(param);
        index++;
      }
      return params;
    }
  }
}

app.get('/article/:author/:slug', (res, req) => {
  // There should be some middleware to extend or pre-validate HttpRequest and HttpResponse
  extendHttpRequest(req)

  const [ author, slug ] = req.getParameters();
  res
    .writeHeader('content-type', 'application/json; charset=utf-8')
    .writeStatus('200')
    .end(`{"author": "${author}", "slug": "${slug}"}`);
})

app.listen(3000, (listenSocket) => {
  if (listenSocket) {
    console.log('Listening to port 3000');
  }
});
```

**Reply by yamiteru** [reply]
@jesusvilla do you realise this is useless? It just loops through parameters and saves them into an array.. but parameters are already an array so there is absolutely no added benefit doing this.

**Reply by jesusvilla** [reply]
@yamiteru Of course it's useless, it even decreases performance, but it's the only way to achieve that new method you guys are talking about, I just wanted to show you that it's the way that exists today to do that, since there is no native way of uws. The parameters are not an array. Also, I mention about the possibility of extending uws.HttpRequest and uws.HttpResponse.

**Reply by yamiteru** [reply]
```javascript
// this is array
const params = [];
// then you return the array
return params;
// and then you destructure the array
const [ author, slug ] = req.getParameters();
```

So yes, parameters are an array.
If you want to access them by name you have to map the array to an object.

**Reply by jesusvilla** [reply]
Yeah, but I mean that doesn't exist yet.

**Comment by a-tonchev** [comment]
Yes it is easy to make own solution. This is mine:

```
app.get('/demoGet/:category/somethingElse/:id', (res, req) => {
  const params = {};

  ['category', 'id'].forEach((param, paramIndex) => {
    params[param] = req.getParameter(paramIndex);
  });

  const { category, id } = params;
  res.writeStatus('200').end(`Category: ${category}, ID: ${id}!`);
});
```

**Reply by yamiteru** [reply]
Here you go a reusable and faster version:

```javascript
const getParameters = (req, keys) => {
  const obj = {};
  const params = req.getParameters();
  const length = params.length;

  for(let i = 0; i < length; ++i) {
    const key = keys[i];
    if(key) obj[key] = params[i];
  }

  return obj;
}
```
